### PR TITLE
New branch protection syntax

### PR DESCRIPTION
**Pull request by soerenmartius**
This is a draft to remove the list from branch_protections and to work with objects only.

### CODE REVIEW

**Comment by soerenmartius** [comment]
So this will still fail in some cases. For example, if we don't provide a complete branch_protection but we don't set the restrictions key:

```
branch_protection_rules = [
    {
      branch                 = "master"
      enforce_admins         = true
      require_signed_commits = false

      required_status_checks = {
        strict = false
        contexts = [
        "ci/semaphoreci"]
      }


      required_pull_request_reviews = {
        dismiss_stale_reviews           = true
        require_code_owner_reviews      = true
        required_approving_review_count = 1
      }

    }
  ]
```

Terraform will throw an exception
```
  on .terraform/modules/terraform-github-organization/main.tf line 94, in resource "github_branch_protection" "branch_protection_rule":
  94:     users = lookup(local.branch_protection_rules[count.index].restrictions, "users", [])
    |----------------
    | count.index is 0
    | local.branch_protection_rules is tuple with 1 element

Invalid value for "inputMap" parameter: lookup() requires a map as the first
argument.
```

What's the recommended workaround @mariux ?

File: main.tf [hunk 1]
```diff
@@ -20,43 +20,25 @@ locals {
 locals {
   branch_protection_rules = [
     for b in var.branch_protection_rules : merge({
-      branch                        = null
-      enforce_admins                = null
-      require_signed_commits        = null
-      required_status_checks        = []
-      required_pull_request_reviews = []
-      restrictions                  = []
```
Comment (mariux) [inline main.tf]: should default to `{}`

File: main.tf [hunk 2]
```diff
@@ -179,9 +149,18 @@ resource "github_repository_collaborator" "collaborator" {
...
+    permission = "admin"
+  }]
+  team_push = [for i in var.push_team_ids : {
+    team_id    = i,
+    permission = "push"
+  }]
+  team_pull = [for i in var.pull_team_ids : {
+    team_id    = i,
+    permission = "pull"
+  }]
```
Comment (mariux) [inline main.tf]: i think it was more readable before... but just an opinion ;)

File: main.tf [hunk 3]
```diff
@@ -107,34 +89,22 @@ resource "github_branch_protection" "branch_protection_rule" {
...
-      require_code_owner_reviews      = required_pull_request_reviews.value.require_code_owner_reviews
-      required_approving_review_count = required_pull_request_reviews.value.required_approving_review_count
-    }
+  required_pull_request_reviews {
+    dismiss_stale_reviews           = lookup(local.branch_protection_rules[count.index].required_pull_request_reviews, "dismiss_stale_reviews", true)
+    dismissal_users                 = lookup(local.branch_protection_rules[count.index].required_pull_request_reviews, "dismissal_users", [])
+    dismissal_teams                 = lookup(local.branch_protection_rules[count.index].required_pull_request_reviews, "dismissal_teams", [])
+    require_code_owner_reviews      = lookup(local.branch_protection_rules[count.index].required_pull_request_reviews, "require_code_owner_reviews", null)
+    required_approving_review_count = lookup(local.branch_protection_rules[count.index].required_pull_request_reviews, "required_approving_review_count", null)
   }
```
Comment (mariux) [inline main.tf]: i would keep the for_each as we do not want to set this when there are no reviews required..
the reference is just getting crazy complicated.. 
```
for_each = length(keys(local.branch_protection_rules[count.index].required_pull_request_reviews)) > 0 ? .....
```

but maybe also make the resource a `for_each` keyed by branch as we the can just refrence `each.value.required_pull_request_reviews` making it more readable..

File: main.tf [hunk 4]
```diff
@@ -107,34 +77,22 @@ resource "github_branch_protection" "branch_protection_rule" {
...
-  dynamic "restrictions" {
-    for_each = local.restrictions[count.index]
-
-    content {
-      users = restrictions.value.users
-      teams = restrictions.value.teams
-    }
+  restrictions {
+    users = lookup(local.branch_protection_rules[count.index].restrictions, "users", [])
+    teams = lookup(local.branch_protection_rules[count.index].restrictions, "teams", [])
```
Comment (mariux) [inline main.tf]: is default here empty array or `null`?

**Comment by mariux** [comment]
closing in favor of #19

**Pull request closed without merging**